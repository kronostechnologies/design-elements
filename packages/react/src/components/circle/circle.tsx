import { VoidFunctionComponent } from 'react';
import styled from 'styled-components';
import { useTheme } from '../../hooks/use-theme';

const Svg = styled.svg`
    display: block;
`;

const BackgroundCircle = styled.circle``;

const FillCircle = styled.circle`
    transform: rotate(90deg);
    transform-origin: 50% 50%;
`;

interface CircleProps {
    color: string;
    percent: number;
    radius: number;
    stroke: number;
}

export const Circle: VoidFunctionComponent<CircleProps> = ({
    color, percent, radius, stroke,
}) => {
    const normalizedRadius = radius - (stroke / 2);
    const circumference = normalizedRadius * 2 * Math.PI;
    const strokeDashoffset = circumference - ((percent / 100) * circumference);
    const theme = useTheme();
    const diameter = radius * 2;

    return (
        <Svg height="100%" width="100%" viewBox={`0 0 ${diameter} ${diameter}`}>
            <BackgroundCircle
                stroke={theme.colors.grey}
                fill="transparent"
                strokeWidth={stroke}
                strokeLinecap="round"
                r={normalizedRadius}
                cx={radius}
                cy={radius}
            />
            <FillCircle
                stroke={color}
                fill="transparent"
                strokeWidth={stroke}
                strokeDasharray={`${circumference} ${circumference}`}
                strokeDashoffset={strokeDashoffset}
                strokeLinecap="round"
                r={normalizedRadius}
                cx={radius}
                cy={radius}
            />
        </Svg>
    );
};
