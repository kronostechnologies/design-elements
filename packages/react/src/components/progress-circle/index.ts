export { ProgressCircle } from './progress-circle';
export type { ProgressCircleProps } from './progress-circle';
