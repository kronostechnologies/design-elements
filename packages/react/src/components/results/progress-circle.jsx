import React, { Component } from "react";
import styled from "styled-components";

// Source: https://css-tricks.com/building-progress-ring-quickly/
class ProgressRing extends React.Component {
  constructor(props) {
    super(props);

    const { radius, stroke } = this.props;

    this.normalizedRadius = radius - stroke * 2;
    this.circumference = this.normalizedRadius * 2 * Math.PI;
  }

  render() {
    const { radius, stroke, progress } = this.props;
    const strokeDashoffset =
      this.circumference - (progress / 100) * this.circumference;
    const PlaceholderDashoffset = this.circumference - 1 * this.circumference;

    const Number = styled.div`
      width: ${props => props.rad * 2}px;
      height: ${props => props.rad * 2}px;
      display: flex;
      justify-content: center;
      align-items: center;
      position: absolute;
      top: 0;
      left: 0;
    `;

    const Label = styled.p`
      width: ${props => props.rad * 2}px;
      text-align: center;
    `;

    const style = {
      position: "relative"
    };

    return (
      <div>
        <div>
          <svg height={radius * 2} width={radius * 2} style={style}>
            <circle
              stroke="#DCDCDC"
              fill="transparent"
              strokeWidth={stroke}
              strokeDasharray={this.circumference + " " + this.circumference}
              style={{ PlaceholderDashoffset }}
              stroke-width={stroke}
              strokeLinecap="round"
              r={this.normalizedRadius}
              cx={radius}
              cy={radius}
            />
            <circle
              stroke={this.props.color}
              fill="transparent"
              strokeWidth={stroke}
              strokeDasharray={this.circumference + " " + this.circumference}
              style={{ strokeDashoffset }}
              stroke-width={stroke}
              strokeLinecap="round"
              r={this.normalizedRadius}
              cx={radius}
              cy={radius}
            />
          </svg>
          <Number rad={radius}>
            <p>{this.props.number}</p>
          </Number>
        </div>
        <Label rad={radius}>{this.props.label}</Label>
      </div>
    );
  }
}

export default ProgressRing;
