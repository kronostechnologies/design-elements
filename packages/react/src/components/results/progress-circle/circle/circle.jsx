import React from 'react';
import styled from 'styled-components';

const Circle = ({ radius, stroke, percent, color }) => {
    const normalizedRadius = radius - (stroke * 2);
    const circumference = normalizedRadius * 2 * Math.PI;
    const placeholderDashoffset = circumference - (1 * circumference);
    const strokeDashoffset = circumference - ((percent / 100) * circumference);

    const CirclePath = styled.circle`
      transform: rotate(90deg);
      transform-origin: 50% 50%;
    `;

    return (
        <svg height={radius * 2} width={radius * 2}>
            <circle
                stroke="#DCDCDC"
                fill="transparent"
                strokeWidth={stroke}
                strokeDasharray={`${circumference} ${circumference}`}
                style={{ placeholderDashoffset }}
                strokeLinecap="round"
                r={normalizedRadius}
                cx={radius}
                cy={radius}
            />
            <CirclePath
                stroke={color}
                fill="transparent"
                strokeWidth={stroke}
                strokeDasharray={`${circumference} ${circumference}`}
                style={{ strokeDashoffset }}
                strokeLinecap="round"
                r={normalizedRadius}
                cx={radius}
                cy={radius}
            />
        </svg>
    );
};

export { Circle };
