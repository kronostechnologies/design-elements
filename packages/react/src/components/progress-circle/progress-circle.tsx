import React, { ReactElement } from 'react';
import styled from 'styled-components';
import { Theme } from '../../themes';
import { Circle } from '../circle/circle';

const RADIUS = 62;
const STROKE = 8;

interface ResultProps {
    theme: Theme;
    secondary?: boolean;
}

const Container = styled.div`
    display: inline-block;
    height: auto;
    width: 100%;
`;

const Wrapper = styled.div`
    height: inherit;
    margin-bottom: var(--spacing-1x);
    position: relative;
    width: inherit;
`;

const Result = styled.div`
    align-items: center;
    display: flex;
    height: 100%;
    justify-content: center;
    left: 0;
    position: absolute;
    top: 0;
    width: 100%;

    p {
        color: ${(props: ResultProps) => props.theme.greys.black};
        font-size: 1.5rem;
        margin: 0;
        padding: 0;
    }
`;

const Label = styled.p`
    color: ${(props: ResultProps) => props.theme.greys.black};
    font-size: 1rem;
    letter-spacing: 0.46px;
    line-height: 1.5rem;
    margin: 0;
    padding: 0;
    text-align: center;
`;

const StyledCircle = styled(Circle)`
    height: auto;
    min-height: 124px;
    width: 100%;
`;

interface Props {
    color: string;
    /** Bottom label */
    descriptionLabel?: string;
    percent: number;
    /** Center label */
    resultLabel: string | number;
}

// Source: https://css-tricks.com/building-progress-ring-quickly/
export function ProgressCircle({
    color, descriptionLabel, percent, resultLabel,
}: Props): ReactElement {
    return (
        <Container>
            <Wrapper>
                <StyledCircle
                    radius={RADIUS}
                    stroke={STROKE}
                    percent={percent}
                    color={color}
                />
                <Result>
                    <p>{resultLabel}</p>
                </Result>
            </Wrapper>
            {descriptionLabel && <Label>{descriptionLabel}</Label>}
        </Container>
    );
}
