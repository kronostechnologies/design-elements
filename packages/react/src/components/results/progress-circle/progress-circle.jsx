import React from "react";
import styled from "styled-components";

import { Circle } from './circle/circle';

// Source: https://css-tricks.com/building-progress-ring-quickly/
const progressCircle = ({ percent, color, label, main, result }) => {
    const radius = 73;
    const stroke = 8;

    // Styled Components
    const Container = styled.div`
        display: inline-block;
    `;

    const Wrapper = styled.div`
        position: relative;
        width: ${(radius * 2) / 16}rem;
        height: ${(radius * 2) / 16}rem;
    `;

    const Number = styled.div`
        width: ${(radius * 2) / 16}rem;
        height: ${(radius * 2) / 16}rem;
        display: flex;
        justify-content: center;
        align-items: center;
        position: absolute;
        top: 0;
        left: 0;
        p {
          font-size: 1.625rem;
        }
    `;

    const Label = styled.p`
        width: ${(radius * 2) / 16}rem;
        text-align: center;
        color: ${props => props.main ? 'rgb(0,0,0)' : "rgb(87,102,110)"};
    `;

    return (
        <Container>
            <Wrapper>
                <Circle radius={radius} stroke={stroke} percent={percent} color={color} />
                <Number>
                    <p>{result}</p>
                </Number>
            </Wrapper>
            <Label main={main}>{label}</Label>
        </Container>
    );
};

export default progressCircle;
