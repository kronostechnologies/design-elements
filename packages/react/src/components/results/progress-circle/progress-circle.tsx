import React, { ReactElement, ReactText } from 'react';
import styled from 'styled-components';

import { Circle } from './circle/circle';

const RADIUS = 73;
const STROKE = 8;

function getColor(secondary: boolean | undefined): string {
    return secondary ? 'rgb(87,102,110)' : 'rgb(0, 0, 0)';
}

interface ResultProps extends Pick<Props, 'secondary'> {
}

const Container = styled.div`
    display: inline-block;
`;

const Wrapper = styled.div`
    height: ${(RADIUS * 2) / 16}rem;
    position: relative;
    width: ${(RADIUS * 2) / 16}rem;
`;

const Result = styled.div`
    align-items: center;
    display: flex;
    height: ${(RADIUS * 2) / 16}rem;
    justify-content: center;
    left: 0;
    position: absolute;
    top: 0;
    width: ${(RADIUS * 2) / 16}rem;
    p {
        color: ${(props: ResultProps) => getColor(props.secondary)};
        font-size: 1.625rem;
    }
`;

const Label = styled.p`
    color: ${(props: ResultProps) => getColor(props.secondary)};
    text-align: center;
    width: ${(RADIUS * 2) / 16}rem;
`;

interface Props {
    color: string;
    descriptionLabel?: string;
    resultLabel: ReactText;
    percent: number;
    secondary?: boolean;
}

// Source: https://css-tricks.com/building-progress-ring-quickly/
export function ProgressCircle({ color, descriptionLabel, percent, resultLabel, secondary }: Props): ReactElement {
    return (
        <Container>
            <Wrapper>
                <Circle
                    radius={RADIUS}
                    stroke={STROKE}
                    percent={percent}
                    color={color}
                />
                <Result secondary={secondary}>
                    <p>{resultLabel}</p>
                </Result>
            </Wrapper>
            {descriptionLabel && <Label secondary={secondary}>{descriptionLabel}</Label>}
        </Container>
    );
}
