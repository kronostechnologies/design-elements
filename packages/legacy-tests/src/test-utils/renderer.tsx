import { CommonWrapper, mount, MountRendererProps, ReactWrapper, render } from 'enzyme';
import { act, Component, FunctionComponent, PropsWithChildren, ReactElement } from 'react';
import { MemoryRouter } from 'react-router-dom';
import { DesignSystem, DesignSystemProps } from '~/components/design-system';
import { ThemeWrapper } from '~/components/theme-wrapper/theme-wrapper';

export const AllProviders: FunctionComponent<PropsWithChildren<DesignSystemProps>> = ({
    children,
    language,
    theme,
    staticDevice,
}) => (
    <MemoryRouter future={{ v7_startTransition: false, v7_relativeSplatPath: false }}>
        <DesignSystem language={language} staticDevice={staticDevice} theme={theme}>
            {children}
        </DesignSystem>
    </MemoryRouter>
);

interface WrappingComponentProps {
    wrappingComponentProps?: DesignSystemProps;
}

type Options = MountRendererProps & WrappingComponentProps;

export function mountWithProviders<C extends Component, P = C['props'], S = C['state']>(
    component: ReactElement<P>,
    options: Options = {},
): ReactWrapper<P, S, C> {
    return mount(component, {
        ...options,
        wrappingComponent: AllProviders as Options['wrappingComponent'],
    });
}

export function mountWithTheme<C extends Component, P = C['props'], S = C['state']>(
    component: ReactElement<P>,
    options?: MountRendererProps,
): ReactWrapper<P, S, C> {
    return mount(component, {
        wrappingComponent: ThemeWrapper as MountRendererProps['wrappingComponent'],
        ...options,
    });
}

export function renderWithTheme(
    component: ReactElement,
): cheerio.Cheerio {
    return render(
        <ThemeWrapper>
            {component}
        </ThemeWrapper>,
    );
}

export async function actUpdate<C extends Component, P = C['props'], S = C['state']>(
    wrapper: CommonWrapper<P, S, C>,
): Promise<void> {
    await act(async () => {
        wrapper.update();
    });
}

export async function actAndWaitForEffects<C extends Component, P = C['props'], S = C['state']>(
    wrapper: ReactWrapper<P, S, C>,
    action: () => void,
): Promise<ReactWrapper<P, S, C>> {
    await act(async () => action());
    await actUpdate(wrapper);
    return wrapper;
}
