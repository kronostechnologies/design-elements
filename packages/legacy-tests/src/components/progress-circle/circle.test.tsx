import { Circle } from '~/components/progress-circle/circle';
import { renderWithTheme } from '../../test-utils/renderer';

describe('Circle', () => {
    it('Matches the snapshot', () => {
        const tree = renderWithTheme(
            <Circle
                radius={73}
                stroke={8}
                percent={55}
                color="#ccc"
            />,
        );

        expect(tree).toMatchSnapshot();
    });
});
