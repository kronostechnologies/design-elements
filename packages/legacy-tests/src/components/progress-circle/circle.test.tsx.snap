// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Circle Matches the snapshot 1`] = `
.c0 {
  display: block;
}

.c1 {
  -webkit-transform: rotate(90deg);
  -ms-transform: rotate(90deg);
  transform: rotate(90deg);
  -webkit-transform-origin: 50% 50%;
  -ms-transform-origin: 50% 50%;
  transform-origin: 50% 50%;
}

<svg
  class="c0"
  height="100%"
  viewBox="0 0 146 146"
  width="100%"
>
  <circle
    class=""
    cx="73"
    cy="73"
    fill="transparent"
    r="69"
    stroke="#DBDEE1"
    stroke-linecap="round"
    stroke-width="8"
  />
  <circle
    class="c1"
    cx="73"
    cy="73"
    fill="transparent"
    r="69"
    stroke="#ccc"
    stroke-dasharray="433.53978619539146 433.53978619539146"
    stroke-dashoffset="195.09290378792613"
    stroke-linecap="round"
    stroke-width="8"
  />
</svg>
`;
